### ISSUE TITLE
[0.5.1] Coercion and value transformation with malli

### ISSUE THREAD

**Issue by piotr-yuxuan** [metosin/reitit#380]
I've been willing to apply a custom value transformation to a json-encoded body request. JSON keys always are strings and default json transformer makes them keyword. I need to coerce them into numbers. Value transformation provided by malli sounds perfect but I find it hard to replicate these examples from malli readme in a reitit context:

``` clojure
(m/decode
  [int? {:math/multiplier 10
         :decode/math {:compile '(fn [schema _]
                                  (let [multiplier (:math/multiplier (m/properties schema))]
                                    (fn [x] (* x multiplier))))}}]
  12
  (mt/transformer {:name :math}))
```
— https://github.com/metosin/malli#value-transformation

A basic example of how to achieve so would be great.

**Comment by piotr-yuxuan** [comment]
Concrete example: how to bake this logic and access it from reitit malli-based coercion?

``` clojure
(def decoders
  {:decode/keyword (fn [x] (if (keyword? x) (Double/parseDouble (name x)) x))
   :decode/string (fn [x] (if (string? x) (Double/parseDouble x) x))})

(malli/decode
  [:map-of
   [:and decoders number?]
   [:and decoders number? [:>= 0]]]
  {10.5 :0.8687473822227199
   "20.2" 1.3924244031817605
   :100.0 "2.1488255192684806"}
  (mt/transformer {:name :keyword}
                  {:name :string}))
```

**Comment by ikitommi** [comment]
The default `reitit-malli` transformer chains are defined here: https://github.com/metosin/reitit/blob/master/modules/reitit-malli/src/reitit/coercion/malli.cljc. You can override the transformers via `reitit.coercion.malli/create`, see [example here](https://github.com/metosin/reitit/blob/master/examples/ring-malli-swagger/src/example/server.clj#L75-L85).

Something like (not tested):

```clj
(require '[reitit.malli.coercion :as malli])
(require '[malli.core :as m])
(require '[malli.transformer :as mt])

(def my-string-transformer
   (mt/transformer
      {:name :math} ;; support named `:math/decode` & `:math/encode`
      (mt/strip-extra-keys-transformer)
      (mt/string-transformer)
      (mt/default-transformer)))

(def coercion
  (malli/create
     (-> malli/default-options (assoc-in [:transformers :string :default] my-string-transformer))
```

You need to override all the transformers from the default map, so there is a lot of boilerplate. To make things easier, we could:

1) add new options to `reitit.malli.coercion/create` to inject customer transformers into all chains
2) add two new steps to all chains by default: `:before` (before alll) and `:after` (after all).

the latter would make the default string-transformer look like:

```clj
(def string-transformer
   (mt/transformer
      {:name :before} ;; support named `:before/decode` & `:before/encode`
      (mt/strip-extra-keys-transformer)
      (mt/string-transformer)
      (mt/default-transformer)
      {:name :after} ;; support named `:after/decode` & `:after/encode`))
```

ideas and comments welcome.

**Comment by piotr-yuxuan** [comment]
TL;DR: @ikitommi would you accept PR about that?

Expanding the initial message, here are attempts to apply custom value transformation with malli from within reitit. The query parameter `:zoned-date-time` is sent as a string, parsed as a `java.time.ZonedDateTime`, processed by domain logic as such, and then sent back as a string.

- Example 1 is actually working, but it looks a bit like boilerplate. It uses two different schemas `ZonedDateTime-object` and `ZonedDateTime-string` which actually represent two faces of the same coin.
- Example 2 is not working. Transformer `:encode/date` is never called. However in malli code (ns `malli.transform`) we can see references about encoders. Perhaps it's a bug from my side, not sure yet.

What if malli coercion would apply after transformer chain `:decode` in middleware `coercion/coerce-request-middleware` and before transformer chain `:encode` transformer chain in middleware `coercion/coerce-response-middleware`? 

Perhaps I'm completely misguided but I might have three remarks:
- The schemas used in the example, for instance `ZonedDateTime-object` use `:fn` and Java objects. They are not compatible with sci, so here the actual function is passed, not a symbol. Perhaps it is a problem and `:fn` in malli in strongly intended for sci-serialisable functions.
- However sometimes it is desirable and highly helpful to coerce request components to Objects. This way of doing things is kind of already present in malli with keywords, it would awesome to extend it to other types.
``` clojure
(malli.core/decode [keyword?] "i-will-be/a-keyword (mt/json-transformer))
```
For instance when we send a ZonedDateTime as a string, I believe it's more idiomatic to seamlessly provide business code with an actual instance of that class. This would also run for Java enums, and so on.
- Moreover, I might be even more misguided, and surely it's more related to malli than to reitit, but in my humble opinion having such encoders and decoders would allow to seamlessly extend malli to validate and coerce not only Clojure-native data structures. Provided transformers can be recursive (they should be if we apply the visitor pattern, shouldn't they?), we could transform, coerce, and validate other data representations such as Avro records. I feel like it would only be a matter of writing the good transformers (for Avro: mainly enum, string, and maps).

### Example 1
``` clojure
(ns example
  (:require
    [malli.transform :as mt]
    [reitit.ring :as ring]
    [reitit.ring.coercion :as coercion]
    [reitit.ring.middleware.parameters :as parameters]
    [reitit.ring.middleware.muuntaja :as muuntaja])
  (:import (java.time ZonedDateTime)))

(def router-opts
  {:data {:coercion (reitit.coercion.malli/create
                      (-> reitit.coercion.malli/default-options
                          (assoc :transformers {:body {:default (mt/transformer {:name :date})
                                                       :formats {"application/json" (mt/transformer mt/json-transformer {:name :date})}}
                                                :string {:default (mt/transformer mt/string-transformer {:name :date})}
                                                :response {:default (mt/transformer {:name :date})}})))
          :middleware [parameters/parameters-middleware
                       muuntaja/format-negotiate-middleware
                       muuntaja/format-response-middleware
                       muuntaja/format-request-middleware
                       coercion/coerce-response-middleware
                       coercion/coerce-request-middleware]}})

(def ZonedDateTime-object
  [:fn {:swagger/description "java.util.ZonedDateTime"
        :swagger/example "2020-05-27T16:14:29.375917+01:00[Europe/Berlin]"
        :swagger/type :string
        :decode/date {:enter #(do (println "ZonedDateTime-object decode" (type %))
                                  (ZonedDateTime/parse %))}}
   #(do (println "ZonedDateTime-object valid?" (type %))
        (instance? ZonedDateTime %))])

(def ZonedDateTime-string
  [:fn {:swagger/description "java.util.ZonedDateTime"
        :swagger/example "2020-05-27T16:14:29.375917+01:00[Europe/Berlin]"
        :swagger/type :string
        :decode/date {:enter #(do (println "ZonedDateTime-string decode" (type %))
                                  (str %))}}
   #(do
      (println "ZonedDateTime-string valid?" (type %))
      (string? %))])

(def handler-object-string
  (ring/ring-handler
    (ring/router
      [["/mwe" {:get {:tags #{"example"}
                      :handler (fn [request]
                                 (let [^ZonedDateTime zdt (-> request :parameters :query :zoned-date-time)]
                                   {:status 200, :body {:zoned-date-time (.plusYears zdt 42)}}))
                      :parameters {:query [:map [:zoned-date-time ZonedDateTime-object]]}
                      :responses {200 {:description ""
                                       :body [:map [:zoned-date-time ZonedDateTime-string]]}}}}]]
      router-opts)))

(handler-object-string
  {:request-method :get
   :uri "/mwe"
   :query-params {:zoned-date-time "2020-05-27T16:14:29.375917+01:00[Europe/Berlin]"}})
;; repl stdout
;; ZonedDateTime-object decode java.lang.String
;; ZonedDateTime-object valid? java.time.ZonedDateTime
;; ZonedDateTime-string decode java.time.ZonedDateTime
;; ZonedDateTime-string valid? java.lang.String
;; ZonedDateTime-string decode java.lang.String
;; ZonedDateTime-string valid? java.lang.String
;; => {:status 200, :body {:zoned-date-time "2062-05-27T17:14:29.375917+02:00[Europe/Berlin]"}}
```

### Example 2
(code continued)
``` clojure
(def ZonedDateTime-unified
  [:fn {:swagger/description "java.util.ZonedDateTime"
        :swagger/example "2020-05-27T16:14:29.375917+01:00[Europe/Berlin]"
        :swagger/type :string
        :decode/date {:enter #(do (println "ZonedDateTime-unified decode" (type %))
                                  (ZonedDateTime/parse %))}
        :encode/date {:enter #(do ;; encode is never called
                                (println "ZonedDateTime-unified encode" (type %))
                                (str %))}}
   #(do (println "ZonedDateTime-unified valid?" (type %))
        (instance? ZonedDateTime %))])

(def handler-unified
  (ring/ring-handler
    (ring/router
      [["/mwe" {:get {:tags #{"example"}
                      :handler (fn [request]
                                 (let [^ZonedDateTime zdt (-> request :parameters :query :zoned-date-time)]
                                   {:status 200, :body {:zoned-date-time (.plusYears zdt 42)}}))
                      :parameters {:query [:map [:zoned-date-time ZonedDateTime-unified]]}
                      :responses {200 {:description ""
                                       :body [:map [:zoned-date-time ZonedDateTime-unified]]}}}}]]
      router-opts)))

(handler-unified
  {:request-method :get
   :uri "/mwe"
   :query-params {:zoned-date-time "2020-05-27T16:14:29.375917+01:00[Europe/Berlin]"}})
;; repl stdout
;; ZonedDateTime-unified decode java.lang.String
;; ZonedDateTime-unified valid? java.time.ZonedDateTime
;; ZonedDateTime-unified decode java.time.ZonedDateTime
;; Execution error (ClassCastException) at example/fn.
;; class java.time.ZonedDateTime cannot be cast to class java.lang.CharSequence
```
